### PR TITLE
Ra dspdc 1636 snapshot check

### PR DESCRIPTION
**Pull request by raaidbroad**
## Why
We recently experienced a workflow run that was marked as successful but didn't result in the desired end state (relevant slack [here](https://broadinstitute.slack.com/archives/CHMQPFCFP/p1615485693010300)). We want to ensure that we don't have any false positives like this. [Relevant ticket](https://broadinstitute.atlassian.net/browse/DSPDC-1636).

## This PR
Adds a check to make sure that the latest release that we have in the xml_archive table (which simply downloads the latest xml archive from ClinVar's FTP server) corresponds to the latest release date in a snapshot name. So if the latest release in the xml_archive table is 2020_03_15, and the latest snapshot release date is 2020_03_08, it will yell and fail.

Specifically:
- Adds a new docker image for `check-snapshot` with the relevant python code to actually do the check
- Adds a step to the `process-and-reingest-release.yaml` workflow to call the check-snapshot code

### CODE REVIEW

File: docker/check-snapshot/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+# Install app dependencies
+COPY requirements.txt ./
+
+RUN pip install -r requirements.txt
+
+# Bundle app source
+COPY check.py /app
+
+ENTRYPOINT ["python"]
```
Comment (aherbst-broad) [inline docker/check-snapshot/Dockerfile:14]: ⛏️ newline

File: docker/check-snapshot/check.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+logging.basicConfig(level=logging.INFO)
+
+ENV = os.getenv("ENV")
+google_project = os.getenv("REPO_DATA_PROJECT")
+data_repo_host = os.getenv("REPO_HOST")
+dataset_name = os.getenv("REPO_DATASET_NAME")
+dataset_id = os.getenv("REPO_DATASET_ID")
+
+
+def default_google_access_token():
```
Comment (aherbst-broad) [inline docker/check-snapshot/check.py:19]: 🌱 I am seeing a common pattern of jade + bq scripts that work off of default google app tokens and the repo client. 🤔 what if we create a shared lib for such things? (out of scope for this PR obviously)

File: docker/check-snapshot/check.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    config.access_token = default_google_access_token()
+    client = ApiClient(configuration=config)
+    client.client_side_validation = False
+
+    return RepositoryApi(api_client=client)
+
+
+
+# get latest row from xml archive: SELECT release_date FROM BLAH ORDER BY release_date DESC LIMIT 1
+def get_latest_xml_release_date(project: str, dataset: str):
```
Comment (aherbst-broad) [inline docker/check-snapshot/check.py:39]: ⛏️ Needs a return type annotation

File: docker/check-snapshot/check.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        SELECT release_date
+        FROM `{project}.{dataset}.xml_archive`
+        ORDER BY release_date DESC LIMIT 1
+    """
+    results = bq_client.query(query)
+    release_date = [row[0] for row in results][0]
+    cleaned_date = str(release_date).replace("-", "_")
+    return cleaned_date
+
+def check_snapshot_exists(host: str, dataset_id: str, filter: str):
```
Comment (aherbst-broad) [inline docker/check-snapshot/check.py:51]: ⛏️ Needs a return type annotation

File: docker/check-snapshot/check.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    jade_client = get_api_client(host=host)
+    r = jade_client.enumerate_snapshots(
+        limit=1,sort="created_date", direction="desc", dataset_ids=[dataset_id], filter=filter)
+    assert r.total == 1, "No snapshot found for latest release date in xml_archive table"
+
+def run():
+    latest_release_date = get_latest_xml_release_date(project=google_project, dataset=dataset_name)
+    logging.info(f"Latest release date in XML archive is {latest_release_date}")
```
Comment (aherbst-broad) [inline docker/check-snapshot/check.py:59]: ❤️ I like logging statements like this, they make debugging so much easier.

File: docker/check-snapshot/check.py (hunk 2 above)
Comment (raaidbroad) [reply docker/check-snapshot/check.py:19]: I've thought about this too; I hope/imagine the need would be reduced once we simply write this code in Python for Dagster, so I didn't want to invest in code for the Argo Workflow way of doing things. If we'd still find it useful to have a library for these things though then yeah we should!

File: docker/check-snapshot/check.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    """
+    results = bq_client.query(query)
+    release_date = [row[0] for row in results][0]
+    cleaned_date = str(release_date).replace("-", "_")
+    return cleaned_date
+
+def check_snapshot_exists(host: str, dataset_id: str, filter: str) -> bool:
+    jade_client = get_api_client(host=host)
+    r = jade_client.enumerate_snapshots(
+        limit=1,sort="created_date", direction="desc", dataset_ids=[dataset_id], filter=filter)
```
Comment (aherbst-broad) [inline docker/check-snapshot/check.py:54]: ❓ what are the semantics of the filter? Is it basically a sql like '%foo%'?

**Review by aherbst-broad — APPROVED** [review]
lgtm

File: docker/check-snapshot/check.py (hunk 6 above)
Comment (raaidbroad) [reply docker/check-snapshot/check.py:54]: it's a string contains filter I believe, here is the endpoint: https://jade-terra.datarepo-prod.broadinstitute.org/swagger-ui.html#/repository/enumerateSnapshots

**Review by quazi-broad — APPROVED** [review]
looks good to me. hooray for giving clinvar some love!

**Pull request merged**